### PR TITLE
Lock pyright version

### PR DESCRIPTION
**Pull request by artificial-aidan**
Took a stab at #23. A couple of caveats.

1. `macos-latest` recently deprecated python 3.6, so we removed that. https://github.com/actions/virtual-environments/issues/4060
2. The `version.py` script got a little messy, would take some input on cleaning that up, or another way to do it
3. To test this, I set the `__pyright_version__` 1 behind, which lets the workflow get triggered and a PR created. See here for an automated PR: https://github.com/artificialinc/pyright-python/pull/9

Open to input on the whole process, but wanted to get something started.

### CODE REVIEW

**Comment by artificial-aidan** [comment]
One other thing.

We specifically grabbed the latest version from the npm registry, as that is where the pyright wrapper pulls it from. Didn't want to trigger off of github releases and have npm be out of sync possibly.

File: dev-requirements.txt
Comment (RobertCraigie) [inline dev-requirements.txt:2]: Updating to a beta version adds type hints which fixes one of the `# type: ignore` comments in `version.py`
```suggestion
semver==3.0.0.dev3
```

File: pyright/cli.py
Comment (RobertCraigie) [inline pyright/cli.py:9]: I prefer grouped imports
```suggestion
from . import __pyright_version__, node
from .utils import env_to_bool
```

File: tests/test_main.py
Comment (RobertCraigie) [inline tests/test_main.py:53]: We should probably check that the version is greater than or equal to the auto-updated pyright version
```suggestion
    assert version.parse(match.group(1)) > version.parse(__pyright_version__)
```

File: version.py
Comment (RobertCraigie) [inline version.py:44]: These `# type: ignore` comments can be removed by switching from `FileInput` to `fileinput.input()`
```diff
- with FileInput('pyright/__init__.py', inplace=True) as f:
+ with fileinput.input('pyright/__init__.py', inplace=True) as f
```

**Review by RobertCraigie — CHANGES_REQUESTED** [review]
Some nitpicks.

Also if you are okay with giving me write access to your fork then I'd be more than happy to just push the changes myself.

File: version.py
Comment (artificial-aidan) [reply version.py:44]: Ah, i knew there had to be a simple way.

File: pyright/cli.py
Comment (artificial-aidan) [reply pyright/cli.py:9]: Agreed, the last projects i was working in had isort enabled, so i have stopped thinking about grouping them myself.

Comment (RobertCraigie) [reply pyright/cli.py:9]: Yeah I should probably enable isort for this project too.

**Comment by RobertCraigie** [comment]
We also need a mechanism to automatically release the package on PyPi.

Do you have any suggestions?

**Comment by artificial-aidan** [comment]
I saw you have the release.py script. I think calling that but let it take in some credentials would work. You would have to add repo credentials that match the ones you use for pypi.

I could give that a shot on our local repo, and local pypi registry.

**Comment by RobertCraigie** [comment]
Yeah it's more how do we automatically trigger that script?

We could have another workflow that runs on every commit to main checking if the package version has been changed and making a release then.

**Comment by artificial-aidan** [comment]
> Yeah it's more how do we automatically trigger that script?
> 
> We could have another workflow that runs on every commit to main checking if the package version has been changed and making a release then.

Ah. I think you could just run it every time a push to main happens. Pypi will reject it if you try and re-publlish an existing package

**Comment by RobertCraigie** [comment]
I think I've found a better solution.

Move all the version updating and setting to it's own file (`pyright/_version.py`) and re-export those variables in `__init__.py`, then we can update the release workflow so that it only runs when `pyright/_version.py` changes.

`_version.py`
```py
__version__ = '0.0.13'
__pyright_version__ = '1.1.223'
```

`__init__.py`
```py
from ._version import (
    __version__ as __version__,
    __pyright_version__ as __pyright_version__,
)
```

You would then need to:
- update the `version.py` script to change `_version.py` instead of `__init__.py`
- update `setup.py` to read the version from `_version.py`

and that should be all you have to change

**Comment by artificial-aidan** [comment]
Ok, moved that stuff around, and added a release workflow that will run after the Test workflow on the main branch.

Couple todos, I'm testing it on our in house repo, so there are some things to change before merging it to your branch. (pypi repository url and runner tags)

Where did we land on matching version to pyright version?

**Comment by artificial-aidan** [comment]
> ```python
> from ._version import (
>     __version__ as __version__,
>     __pyright_version__ as __pyright_version__,
> )
> ```

Codacy reports the `__version__` import as unused. Should I remove that one?

**Comment by RobertCraigie** [comment]
> Codacy reports the __version__ import as unused. Should I remove that one?

No I will just tell codacy to ignore the issue. I honestly do not know why I have it enabled for this repo.

**Comment by RobertCraigie** [comment]
> Where did we land on matching version to pyright version?

I think matching versions is a good idea.

**Comment by RobertCraigie** [comment]
Let me know when it's ready for review :)

**Comment by artificial-aidan** [comment]
Ok. So I think this PR is ready for a final review.

Actions needed:
1. Repository secrets for PyPi login. (PYPI_USER, PYPI_PASS)
2. Merge to main and do a manual workflow run to generate a PR?

Right now I set the version to `0.0.13.post0` and `1.1.223` so it will first publish `0.0.13.post0` and then it should generate a PR for version `1.1.224` of the package when it gets run on the cronjob or manually

File: version.py
Comment (RobertCraigie) [inline version.py:47]: This appears to not be used anymore

File: dev-requirements.txt
Comment (RobertCraigie) [inline dev-requirements.txt:2]: semver is no longer needed as `bump_pyright_package_ver` is not used.

**Review by RobertCraigie — CHANGES_REQUESTED** [review]
Minor nitpicks, thank you! :)

**Review by RobertCraigie — APPROVED** [review]

**Comment by RobertCraigie** [comment]
Thanks again!

**Comment by artificial-aidan** [comment]
@RobertCraigie I'm pretty sure the combination of `workflow_run` and `paths` isn't working. I thought I had tested it but maybe it isn't supported.

I was trying to reuse workflows, but it might be easier to just copy over the test.yml steps in to the release.yml, and have it run on a push to main only if `pyright/_version.py` changes. Basically just have the test job defined in release.yml and test.yml, and make the `publish_release` job require the test job to run first.

**Comment by RobertCraigie** [comment]
@artificial-aidan That is unfortunate. In that case I think I'll just remove `workflow_run` from `release.yml` as all releases will be made through a PR which I will not merge unless tests pass anyway.

Thanks for the heads up.

**Pull request merged**
